lib/Frontend/ArgsToFrontendOutputsConverter.cpp

```diff
@@ -202,3 +218,225 @@ OutputFilesComputer::deriveOutputFileFromParts(StringRef dir,
+
+  if (!serializedDiagnosticsPath)
+    return None;
+
+  auto objcHeaderOutputPath = determineSupplementaryOutputFilename(
+      OPT_emit_objc_header, pathsFromArguments.ObjCHeaderOutputPath, "h", "",
+      defaultSupplementaryOutputPathExcludingExtension);
+  if (!objcHeaderOutputPath)
+    return None;
+
+  auto loadedModuleTracePath = determineSupplementaryOutputFilename(
+      OPT_emit_loaded_module_trace, pathsFromArguments.LoadedModuleTracePath,
+      "trace.json", "", defaultSupplementaryOutputPathExcludingExtension);
+  if (!loadedModuleTracePath)
+    return None;
+
+  auto tbdPath = determineSupplementaryOutputFilename(
+      OPT_emit_tbd, pathsFromArguments.TBDPath, "tbd", "",
+      defaultSupplementaryOutputPathExcludingExtension);
+  if (!tbdPath)
+    return None;
+
+  auto moduleDocOutputPath = determineSupplementaryOutputFilename(
+      OPT_emit_module_doc, pathsFromArguments.ModuleDocOutputPath,
+      SERIALIZED_MODULE_DOC_EXTENSION, "",
+      defaultSupplementaryOutputPathExcludingExtension);
+  if (!moduleDocOutputPath)
+    return None;
+
+  ID emitModuleOption;
+  std::string moduleExtension;
+  std::string mainOutputIfUsableForModule;
+  deriveModulePathParameters(emitModuleOption, moduleExtension,
+                             mainOutputIfUsableForModule);
+
+  auto moduleOutputPath = determineSupplementaryOutputFilename(
+      emitModuleOption, pathsFromArguments.ModuleOutputPath, moduleExtension,
+      mainOutputIfUsableForModule,
+      defaultSupplementaryOutputPathExcludingExtension);
+  if (!moduleOutputPath)
+    return None;
+
+  SupplementaryOutputPaths sop;
+  sop.ObjCHeaderOutputPath = *objcHeaderOutputPath;
+  sop.ModuleOutputPath = *moduleOutputPath;
+  sop.ModuleDocOutputPath = *moduleDocOutputPath;
+  sop.DependenciesFilePath = *dependenciesFilePath;
+  sop.ReferenceDependenciesFilePath = *referenceDependenciesFilePath;
+  sop.SerializedDiagnosticsPath = *serializedDiagnosticsPath;
+  sop.LoadedModuleTracePath = *loadedModuleTracePath;
+  sop.TBDPath = *tbdPath;
+  return sop;
+}
+
+StringRef SupplementaryOutputPathsComputer::
+    deriveDefaultSupplementaryOutputPathExcludingExtension(
+        StringRef outputFilename, const InputFile &input) const {
+  // Put the serialized diagnostics file next to the output file if possible.
+  if (!outputFilename.empty() && outputFilename != "-")
+    return outputFilename;
+
+  if (input.isPrimary() && input.file() != "-")
+    return llvm::sys::path::filename(input.file());
+
+  return ModuleName;
+}
+
+Optional<std::string>
+SupplementaryOutputPathsComputer::determineSupplementaryOutputFilename(
+    options::ID emitOpt, std::string pathFromArguments, StringRef extension,
+    StringRef mainOutputIfUsable,
+    StringRef defaultSupplementaryOutputPathExcludingExtension) const {
+  using namespace options;
+
+  if (!pathFromArguments.empty())
+    return pathFromArguments;
+
+  if (!Args.hasArg(emitOpt))
+    return std::string();
```

Comment:
Shouldn't this be `None`?

Reply:
std::string() is correct, but the code was unclear because it was using None for error. However,
there is no error condition in this version. So, I modified this function and its caller to just return a string, instead of an optional. And, I added a Doxygen comment. (The optional may come back in a future PR, but I'll comment it if so.)